diff --git a/var/spack/repos/builtin/packages/paraview/package.py b/var/spack/repos/builtin/packages/paraview/package.py
index 37cb994ab..f8640cd41 100644
--- a/var/spack/repos/builtin/packages/paraview/package.py
+++ b/var/spack/repos/builtin/packages/paraview/package.py
@@ -14,6 +14,13 @@ class Paraview(CMakePackage):
     homepage = 'http://www.paraview.org'
     url      = "http://www.paraview.org/files/v5.3/ParaView-v5.3.0.tar.gz"
     _urlfmt  = 'http://www.paraview.org/files/v{0}/ParaView-v{1}{2}.tar.gz'
+    generator = 'Ninja'
+
+    version('master',
+            git='https://gitlab.kitware.com/paraview/paraview.git',
+            branch='master',
+            submodules=True,
+            preferred=True)
 
     version('5.6.0', sha256='cb8c4d752ad9805c74b4a08f8ae6e83402c3f11e38b274dba171b99bb6ac2460')
     version('5.5.2', '7eb93c31a1e5deb7098c3b4275e53a4a')
@@ -66,17 +73,18 @@ class Paraview(CMakePackage):
     # depends_on('sqlite') # external version not supported
     depends_on('zlib')
     depends_on('cmake@3.3:', type='build')
+    depends_on('ninja', type='build')
 
-    patch('stl-reader-pv440.patch', when='@4.4.0')
+    # patch('stl-reader-pv440.patch', when='@4.4.0')
 
     # Broken gcc-detection - improved in 5.1.0, redundant later
-    patch('gcc-compiler-pv501.patch', when='@:5.0.1')
+    # patch('gcc-compiler-pv501.patch', when='@:5.0.1')
 
     # Broken installation (ui_pqExportStateWizard.h) - fixed in 5.2.0
-    patch('ui_pqExportStateWizard.patch', when='@:5.1.2')
+    # patch('ui_pqExportStateWizard.patch', when='@:5.1.2')
 
     # Broken vtk-m config. Upstream catalyst changes
-    patch('vtkm-catalyst-pv551.patch', when='@5.5.0:5.5.2')
+    # patch('vtkm-catalyst-pv551.patch', when='@5.5.0:5.5.2')
 
     def url_for_version(self, version):
         """Handle ParaView version-based custom URLs."""
@@ -175,6 +183,7 @@ class Paraview(CMakePackage):
         if '+python' in spec:
             cmake_args.extend([
                 '-DPARAVIEW_ENABLE_PYTHON:BOOL=ON',
+                '-DPARAVIEW_PYTHON_VERSION:STRING=2',
                 '-DPYTHON_EXECUTABLE:FILEPATH=%s' % spec['python'].command.path
             ])
 
