diff --git a/vtkm/cont/cuda/internal/DeviceAdapterTagCuda.h b/vtkm/cont/cuda/internal/DeviceAdapterTagCuda.h
index 60dbda1b..4d395abe 100644
--- a/vtkm/cont/cuda/internal/DeviceAdapterTagCuda.h
+++ b/vtkm/cont/cuda/internal/DeviceAdapterTagCuda.h
@@ -18,7 +18,8 @@
 #if defined(VTKM_CUDA) && defined(VTKM_ENABLE_CUDA)
 VTKM_VALID_DEVICE_ADAPTER(Cuda, VTKM_DEVICE_ADAPTER_CUDA);
 #elif defined(VTKM_ENABLE_CUDA) && !defined(VTKM_NO_ERROR_ON_MIXED_CUDA_CXX_TAG)
-#error When VTK-m is build with CUDA enabled all compilation units that include DeviceAdapterTagCuda must use the cuda compiler
+// allow responsible adults to use VTK-m downstream w/o a cuda compiler
+// # error When VTK-m is build with CUDA enabled all compilation units that include DeviceAdapterTagCuda must use the cuda compiler
 #else
 VTKM_INVALID_DEVICE_ADAPTER(Cuda, VTKM_DEVICE_ADAPTER_CUDA);
 #endif
