--- vtkm/exec/cuda/internal/WrappedOperators.h	2025-04-29 08:30:18.471616022 +0200
+++ /local/apps/vtk-m/vtkm/exec/cuda/internal/WrappedOperators.h	2024-08-20 11:22:29.754010727 +0200
@@ -197,11 +197,19 @@
 // the binary functor is commutative and the T type is is_arithmetic
 //
 //
+#if THRUST_VERSION >= 200500
+template <typename T, typename F>
+struct is_commutative<vtkm::exec::cuda::internal::WrappedBinaryOperator<T, F>>
+  : public ::cuda::std::is_arithmetic<T>
+{
+};
+#else
 template <typename T, typename F>
 struct is_commutative<vtkm::exec::cuda::internal::WrappedBinaryOperator<T, F>>
   : public thrust::detail::is_arithmetic<T>
 {
 };
+#endif
 }
 } //namespace thrust::detail
 
