diff --git a/data b/data
index d4b7ad29b..c77cfc15d 160000
--- a/data
+++ b/data
@@ -1 +1 @@
-Subproject commit d4b7ad29bc56f92fc7f9cb115f54664f3c313646
+Subproject commit c77cfc15d68c437d0ccd868df8187bc92adcbeb1
diff --git a/viskores/filter/contour/worklet/Clip.h b/viskores/filter/contour/worklet/Clip.h
index 370f3fb32..7b01e5647 100644
--- a/viskores/filter/contour/worklet/Clip.h
+++ b/viskores/filter/contour/worklet/Clip.h
@@ -496,7 +496,7 @@ public:
           {
             cellMapOutputToInput.Set(cellsOffset, cellId);
             shapes.Set(cellsOffset, static_cast<viskores::UInt8>(shape.Id));
-            offsets.Set(cellsOffset, cellIndicesOffset);
+            offsets.Set(cellsOffset++, cellIndicesOffset);
             for (viskores::IdComponent pointId = 0; pointId < points.GetNumberOfComponents();
                  ++pointId)
             {
